crates/consensus/src/transaction/rlp.rs

```diff
@@ -4,9 +4,229 @@ use alloy_eips::eip2718::{Eip2718Error, Eip2718Result};
+        }
+
+        let remaining = buf.len();
+        let tx = Self::rlp_decode_fields(buf)?;
+        let signature = Signature::decode_rlp_vrs(buf, bool::decode)?;
+
+        if buf.len() + header.payload_length != remaining {
+            return Err(alloy_rlp::Error::ListLengthMismatch {
+                expected: header.payload_length,
+                got: remaining - buf.len(),
+            });
+        }
+
+        Ok((tx, signature))
+    }
+
+    /// Decodes the transaction from RLP bytes, including the signature
+    /// Produces a [`Signed`].
+    fn rlp_decode_signed(buf: &mut &[u8]) -> alloy_rlp::Result<Signed<Self>> {
+        Self::rlp_decode_with_signature(buf).map(|(tx, signature)| tx.into_signed(signature))
+    }
+
+    /// Decodes the transaction from eip2718 bytes, expecting the given type
+    /// flag.
+    fn eip2718_decode_with_type(buf: &mut &[u8], ty: u8) -> Eip2718Result<Signed<Self>> {
+        let original_buf = *buf;
+
+        if buf.remaining() < 1 {
+            return Err(alloy_rlp::Error::InputTooShort.into());
+        }
+        let actual = buf.get_u8();
+        if actual != ty {
+            return Err(Eip2718Error::UnexpectedType(actual));
+        }
+
+        // OPT: We avoid re-serializing by calculating the hash directly
+        // from the original buffer contents.
+        let (tx, signature) = Self::rlp_decode_with_signature(buf)?;
+        let total_len = tx.eip2718_encoded_length(&signature);
+        let hash = keccak256(&original_buf[..total_len]);
+
+        Ok(Signed::new_unchecked(tx, signature, hash))
+    }
+
+    /// Decodes the transaction from eip2718 bytes, expecting the default type
+    /// flag.
+    fn eip2718_decode(buf: &mut &[u8]) -> Eip2718Result<Signed<Self>> {
+        Self::eip2718_decode_with_type(buf, Self::DEFAULT_TX_TYPE)
+    }
+
+    /// Decodes the transaction from network bytes.
+    fn network_decode_with_type(buf: &mut &[u8], ty: u8) -> Eip2718Result<Signed<Self>> {
+        let header = Header::decode(buf)?;
+        if header.list {
+            return Err(alloy_rlp::Error::UnexpectedList.into());
+        }
+
+        let remaining = buf.len();
+        let res = Self::eip2718_decode_with_type(buf, ty)?;
+
+        if buf.len() + header.payload_length != remaining {
+            return Err(alloy_rlp::Error::UnexpectedLength.into());
+        }
+
+        Ok(res)
+    }
+
+    /// Decodes the transaction from network bytes, expecting the default type
+    /// flag.
+    fn network_decode(buf: &mut &[u8]) -> Eip2718Result<Signed<Self>> {
+        Self::network_decode_with_type(buf, Self::DEFAULT_TX_TYPE)
+    }
+}
+
 /// Helper trait for managing RLP encoding of transactions inside 2718
 /// envelopes.
 #[doc(hidden)]
+#[deprecated(
+    note = "Use RlpEcdsaEncodableTx for encoding, RlpEcdsaDecodableTx for decoding, and RlpTxHash for tx hashing functionality"
+)]
 pub trait RlpEcdsaTx: SignableTransaction<Signature> + Sized {
```

Comment:
can we change this so that this is encodable + decodable and add a blanket impl?